java-checks/src/main/java/org/sonar/java/checks/AbstractPackageInfoChecker.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.java.AnalysisException;
+import org.sonar.java.EndOfAnalysisCheck;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
+import org.sonar.plugins.java.api.InputFileScannerContext;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.PackageDeclarationTree;
+
+public abstract class AbstractPackageInfoChecker implements JavaFileScanner, EndOfAnalysisCheck {
+  private static final Logger LOG = Loggers.get(AbstractPackageInfoChecker.class);
+  private static final String CACHE_KEY_PREFIX = "java:S1228;S4032:package:";
+
+  private static String cacheKey(InputFile inputFile) {
+    return CACHE_KEY_PREFIX + inputFile.key();
+  }
+
+  protected abstract void processFile(InputFileScannerContext context, String packageName);
+
+  @Override
+  public boolean scanWithoutParsing(InputFileScannerContext inputFileScannerContext) {
+    return getPackageFromCache(inputFileScannerContext).map(packageName -> {
+      processFileAndCacheIfApplicable(inputFileScannerContext, packageName);
+      return true;
+    }).orElse(false);
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    PackageDeclarationTree packageDeclaration = context.getTree().packageDeclaration();
+    String packageName = packageDeclaration == null ? null : ExpressionsHelper.concatenate(packageDeclaration.packageName());
+
+    processFileAndCacheIfApplicable(context, packageName);
+  }
+
+  private void processFileAndCacheIfApplicable(InputFileScannerContext context, @Nullable String packageName) {
+    if (context.getCacheContext().isCacheEnabled()) {
+      writePackageNameToCache(context, packageName == null ? "" : packageName);
+    }
+
+    if (packageName == null || packageName.equals("")) {
+      // default package
+      return;
+    }
+
+    processFile(context, packageName);
+  }
+
+  protected static Optional<String> getPackageFromCache(InputFileScannerContext inputFileScannerContext) {
+    var cacheKey = cacheKey(inputFileScannerContext.getInputFile());
+    try (var in = inputFileScannerContext.getCacheContext().getReadCache().read(cacheKey)) {
```

Comment:
What a strange way of reading from a cache: `InputStream read(String key)`. It means the `read` method has to create an exception with its stack trace just to indicate that the cache does not contain this key...
When a project is not analyzed incrementally, this rule will generate a lot of exceptions and a lot of `LOG.debug` messages even if there are no analyzer errors. And I see only complexity and disadvantages when I see an `InputStream` instead of a `byte[]`.